DietPi-Software | ADS-B Feeder: track airplanes using SDRs and feed the data to ADS-B aggregators

dirkhh opened the pull request:
DietPi-Software | add ADSB-Feeder app that allows feeding ADSB data from one (or more) SDRs to various aggregators (based on https://adsb.im)

Please consider this a draft - I would love feedback on all aspects of this. Did I pick the right category? Any issues with the code? I tried to follow the guidelines I could fine and the comments that you made in our discussions today. But as this is my first time adding code to DietPi, I'm sure there are things that I missed.

This has been tested both in a VM and on an RPi and appears to work ok. I assume there are features in the Web UI that will need tweaking since some things work differently when installed like this.

<!--
Before submitting a pull request:
- Please ensure the target branch is "dev" (active development): https://github.com/MichaIng/DietPi/tree/dev
- Please ensure changes have been tested and verified functional.
-->

dirkhh committed:
add ADSB-Feeder

DietPi-Software | add ADSB-Feeder app that allows feeding ADSB data from
one (or more) SDRs to various aggregators (based on https://adsb.im)

Signed-off-by: Dirk Hohndel <dirk@hohndel.org>

dirkhh committed:
use the adsb-docker service to get restarted with the docker service

With a small change in the upstream repo, this now allows the the
containers to be cleanly stopped / restarted when the same happens to
the docker service, e.g. when the user is installing more software or
runs the benchmarks.

Signed-off-by: Dirk Hohndel <dirk@hohndel.org>

dirkhh committed:
ensure that the ADSB services get correctly handled by DietPi

It turns out that one 'simply has to know' that when adding a service in
dietpi-software, one also has to make sure that the service is handled
in dietpi-services. That took me a while to figure out. But with this it
appears the restarts are handled correctly.

Signed-off-by: Dirk Hohndel <dirk@hohndel.org>

MichaIng commented on dietpi/dietpi-services:
```diff
@@ -272,6 +272,7 @@ _EOF_
 
 		# Additional services: https://github.com/MichaIng/DietPi/issues/1869#issuecomment-401017251
 		[[ -f '/etc/rsyncd.conf' ]] && aSERVICE_NAME+=('rsync')
+		[[ -d '/opt/adsb' ]] && aSERVICE_NAME+=('adsb-docker' 'adsb-setup')
```
These can be added to the array above below YaCy. It is internally checked for the existence of those services and I guess the names are sufficiently specific that we do not need to check for the existence of the install dir as well.

dirkhh replied:
Excellent. I wasn't sure what the preferred way was for doing that as I saw how rsync was added there... I'll make that change.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -4170,6 +4177,50 @@ WantedBy=multi-user.target
+
+			# last but not least, register the services needed with DietPi
+			aENABLE_SERVICES+=('adsb-setup' 'adsb-nonimage' 'adsb-docker')
+			aSTART_SERVICES+=('adsb-setup' 'adsb-nonimage' 'adsb-docker')
```
As the services are added to `dietpi-services`, and as such started/stopped automatically on `dietpi-software` executions, they do not need to be defined here.
```suggestion
```
`adsb-nonimage.service` is started via `BintTo` or similar with one of the other two? Else it could be added to `dietpi-services` as well, ordered in the order of which they need to start, in case one depends on the other.

dirkhh replied:
the `adsb-nonimage.service` needs to run only once, and it must be guaranteed to run first. I'll look into understanding the interdependencies of these different table some more. I'm still quite new to the codebase and saying that plain bash is not my preferred programming environment would be an understatement 🤣

MichaIng replied:
Hmm, currently it runs after `adsb-setup.service`. Since `dietpi-services start` runs before the `aSTART_SERVICES` array, currently `adsb-nonimage.service` would probably need to start right within the install block. If it only needs to run once, is it worth to have it as a service? Or do you mean with "once" that it is a oneshot task (i.e. exits when done) that needs to finish once every time the ADSB feeder is started?

dirkhh replied:
it's a service because that way (at least in theory) I can ensure that it runs before adsb-setup (simply with the `Before:` and `After:` tags in the unit file), and because having something as a service makes it much easier to give it a controlled lifecycle. I have seen some of your other comments on just how broken `systemd` is and how much it is not understanding the needs of SBCs... and having worked with some of the people behind `systemd` for 25 years or so, I hear you and I feel you. All that said, it does do a decent job to deal with those kind of tasks.

And yes, the `-nonimage` service really does only run once. I guess I could manually start it right after the install and never enable it... hmm... food for thought

MichaIng replied:
Haha, that comment was caused by the recent conversation I had with Lennart regarding this particular dbus/logind topic 😄.

Generally I am very happy with systemd, I love the easiness it adds to service management, sandboxing, ordering/dependencies etc. I wished it was more modular, e.g. that one can install the core init system and service management without e.g. systemd-networkd/resolved which we never use and which Debian itself never uses (only Ubuntu uses it OOTB), but I got used to that since the beginning.

But I get annoyed when multiple services/features which currently can be separately installed (`dbus` is a dedicated non-dependency package) are tied into each other and become mandatory (or loudly complain via error message when not present), even that there are trivially to understand use cases (IMO) which never use those features.

More features are not necessarily bad, and I love the alternatives the systemd project provides for a lot of tasks apart of init/service management, from systemd-boot til systemd-container. But the more features are added, the more important it is to keep things module and optional, IMO, so that not everyone is forced to use the whole package with 90% unused bloat, but can select exactly the needed tools/features from systemd (without the need to recompile it). But yeah, nothing is perfect.

_And to be true, we have to look at our own noses with DietPi, which also ties together a bunch of feature 😉._

dirkhh committed:
move around service activation

as requested by @MichaIng.

Signed-off-by: Dirk Hohndel <dirk@hohndel.org>

dirkhh committed:
more complete uninstall for ADSB feeder

Signed-off-by: Dirk Hohndel <dirk@hohndel.org>

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -1591,6 +1591,13 @@ Available commands:
 		aSOFTWARE_DEPS[$software_id]='196'
 		#------------------
+		software_id=141
+		aSOFTWARE_NAME[$software_id]='adsb-feeder'
```
What is your favourite beautified name for this? "ADSB-Feeder", "ADSB Feeder" or "ADS-B Feeder"? 😄
I like to have it consistent everywhere in our scripts and docs and aligned with the preferred spelling of upstream dev(s).

dirkhh replied:
ADS-B Feeder

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -4166,6 +4173,55 @@ WantedBy=multi-user.target
+		then
+			# create the userdata directory and clone the adsb-feeder repo into /tmp
+			G_EXEC mkdir -p /mnt/dietpi_userdata/adsb-feeder
+			G_EXEC_OUTPUT=1 G_EXEC git clone -b dietpi 'https://github.com/dirkhh/adsb-feeder-image.git' /tmp/adsb-feeder
```
As Git is only used on install here, can we just download the archive and skip Git as dependency?
```suggestion
			Download_Install 'https://github.com/dirkhh/adsb-feeder-image/archive/dietpi.tar.gz' /tmp
```
Should download and extract it to `/tmp/adsb-feeder-image-dietpi` (appending the branch) and removes the archive automatically.

Ah, now I see you derive the version string via Git. Okay, if this cannot be extracted easily elsewhere, not a big problem.

dirkhh replied:
yes - that's part of making it possible for people to ask questions and me to know which exact version they are using, so I'd really love to keep the git checkout.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -4166,6 +4173,55 @@ WantedBy=multi-user.target
+
+			# last but not least, register the services needed with DietPi
+			aENABLE_SERVICES+=('adsb-setup' 'adsb-docker')
+			aSTART_SERVICES+=('adsb-setup' 'adsb-docker')
```
```suggestion
```
Not needed anymore thanks to `dietpi-services` 😉.

dirkhh replied:
I thought I removed that already. Oops.

MichaIng committed:
v8.22

- META | Add ADSB Feeder to changelog, readme and dietpi-survey report page

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -4166,6 +4173,55 @@ WantedBy=multi-user.target
+
+			# create the target directory for the app and populated with the code from the git checkout
+			G_EXEC mkdir -p /opt
+			G_EXEC cp -a /tmp/adsb-feeder/src/modules/adsb-feeder/filesystem/root/opt/adsb /opt
```
`/opt` is pre-generated as failsafe step on every install run. But if that directory already exists, I think we should remove it, i.e. doing a clean reinstall? As you have an update service, a reinstall only makes sense when it is really broken, and then it makes sense to make it as clean as possible, excluding only data and configs.
```suggestion
			[[ -d '/opt/adsb' ]] && G_EXEC rm -R /opt/adsb
			G_EXEC cp -a /tmp/adsb-feeder/src/modules/adsb-feeder/filesystem/root/opt/adsb /opt
```
As I see it, can we use `/opt/adsb-feeder` and `/opt/adsb-feeder-update`, to have them unique and a little aligned? I see the latter is generated by `adsb-feeder-update.service`, so it is okay to leave it as it is if this causes you too much hassle/divergence between branches in your code.

dirkhh replied:
I can make that change... that requires making changes both in my adsb-feeder-image upstream repo and here, but I can do that today. I should have plenty of time to wrestle with that (famous last words)

dirkhh replied:
actually... somewhat. `/opt/adsb` is really well established as the folder to use for this setup. Almost all of the documentation that enthusiasts might use will point to that. I would not be thrilled to have to rename that one.
But I did rename `/opt/feeder-update` to `/opt/adsb-feeder-update`
So now they both start with `/opt/adsb`

MichaIng replied:
Okay should be sufficiently unique.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -4166,6 +4173,55 @@ WantedBy=multi-user.target
+			echo "$ADSB_FEEDER_VERSION" > adsb.im.version
+
+			# get the Python module
+			G_AGI python3-flask
```
Not sure whether it is a benefit, just the idea of installing an up-to-date flask via `pip`:
```suggestion
			G_EXEC_OUTPUT=1 G_EXEC pip3 install -U flask
```
Downside is that it is not updated along with `apt upgrade` (security patches), and theoretically a breaking change in flask could happen. Not sure how stable its API is, whether to worry about it.

... ah and via `pip` on 32-bit ARM and RISC-V it might add additional build dependencies, so probably overall not worth it.

dirkhh replied:
It's an interesting question, though. I had one case with old versions of flask included by an OS actually broke my build, but anything that's relatively current I prefer the OS included one as that gets updated and receives patches.

Are all your builds on bookworm at this point?

MichaIng replied:
New Bullseye images are also still provided and Buster systems are still supported. Hence oldest is flask 1.0.2.

dirkhh replied:
ok. that sucks.
I need to work around that then.

MichaIng replied:
We could also disable/hide the software option in Buster and Bullseye (flask 1.1.2), or use `pip` only on those, in case with a fixed/compatible version, like `^2.2.2` to match the one on Bookworm.

If any, I guess flack 1.x vs 2.x is causing issues?

EDIT: Ah, you did that already 👍.

dirkhh replied:
I'm trying to please.
Actually, I'm trying to get this ready to be merged before I'm heading on a trip tomorrow.

dirkhh committed:
make requested changes

Signed-off-by: Dirk Hohndel <dirk@hohndel.org>

dirkhh committed:
get appropriate flask version for older distros

Signed-off-by: Dirk Hohndel <dirk@hohndel.org>

dirkhh committed:
rename the feeder update staging directory

I have the corresponding commit now in
https://github.com/dirkhh/adsb-feeder-image/commit/93efcf3fd5994fc5e3fca5a96f40831bc63c4196

As suggested by @MichaIng.

Signed-off-by: Dirk Hohndel <dirk@hohndel.org>

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -4166,6 +4173,57 @@ WantedBy=multi-user.target
 _EOF_
 		fi
 
+		if To_Install 141 adsb-feeder # ADSB Feeder
```
```suggestion
		if To_Install 141 adsb-setup adsb-docker # ADSB Feeder
```
That's the service(s) which shall be enabled. All enabled services are started at the end automatically, but on a fresh install, they are not automatically enabled (unless it is done via APT package `postinst` script).

dirkhh committed:
install adsb feeder later and fix the services being enabled

Signed-off-by: Dirk Hohndel <dirk@hohndel.org>

MichaIng committed:
v8.22

- DietPi-Software | ADS-B Feeder: Rename from "ADSB Feeder"
- DietPi-Software | ADS-B Feeder: Minor coding: Remove doubled data dir creation, mv instead of cp from cloned reop, make variables local, merge "head" and "awk" calls, preferring "mawk", use single quotes where possible, don't allow removals to fail for non-existent dirs if we definitely expect them, really allow "down" command on uninstall to fail and do not try to remove the Docker image, if Docker is not installed anymore.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -11935,6 +11942,56 @@ _EOF_
+
+			# finally ensure that /run allows executables (this is needed for the containers to be able to not do excessive
+			# writes to physical storage - which might be an SD card)
+			G_EXEC mount -o remount,exec /run
```
Last question: This is really only needed once, for `adsb-nonimage.service` or so? Just because it is not boot persistent and also intended to not be so. Actually using a different tmpfs might be even better to keep `/run` non-executable at all time.

dirkhh replied:
I'm on a plane right now, tracking my own flight using the ADS-B Feeder on DietPi on an RPi3a.
Definitely getting a fair amount of testing.

The `/run` executable is always needed. And after a reboot the service manages it if necessary.
If you would prefer a different tempfs I can of course switch, but there is really no security benefit from having `/run` `noexec` (and most distros don't do that - DietPi is one of the few that does)

MichaIng replied:
The `noexec` is not coming from DietPi, but from Debian, or the responsible Debian package. We do not touch its mount options. But I'm also not sure whether this is a real security benefit, respectively whether anything is relying on it and would cause a security issue with `noexec` removed.

However, it is okay for me to check back about this later and merge this PR. Wishing a pleasant flight, and again, many thanks for this great implementation 🙂!

MichaIng reviewed (approved):
Did some minor coding changes, see my last commit. Tested on all distro versions (VMs) and works pretty well. Ports of the different proxies are also nicely aligned among 109x.

Ready for merging from my end.

Linking the PR for adding this to our docs for completeness: https://github.com/MichaIng/DietPi-Docs/pull/922

dirkhh commented on dietpi/dietpi-software:
```diff
@@ -11942,37 +11942,36 @@ _EOF_
-			echo "ADSB Feeder app running on DietPi" > feeder-image.name
-			echo "$ADSB_FEEDER_VERSION" > adsb.im.version
+			G_EXEC eval 'echo '\''ADSB Feeder app running on DietPi'\'' > feeder-image.name'
+			G_EXEC eval "echo '$ADSB_FEEDER_VERSION' > adsb.im.version"
```
does this work? I thought variables don't get expanded inside single quotes?
I may be out of my shell programming depth here. Not sure why you'd switch this to the eval construct in the first place

MichaIng replied:
As the single quotes are wrapped within double quotes, it does work 🙂. `eval` gets the string with the variable expanded already, so what is between the single quotes then can/must be taken literal.

The whole change here btw is only to make at best really all steps transparent to interested users. Generally I do not like `eval` that much, but I did not find another good way to error-handle writes to files nicely, and the variable is controlled and expanded only once (as of the single quotes).
